EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -78,40 +81,88 @@ public static string ComputePEPValuesForAllPSMsGeneric(List<PeptideSpectralMatch
+            Parallel.ForEach(Partitioner.Create(0, psms.Count),
+                new ParallelOptions { MaxDegreeOfParallelism = maxThreads },
+                (range, loopState) =>
                 {
-                    List<int> indiciesOfPeptidesToRemove = new List<int>();
-                    List<(int notch, PeptideWithSetModifications pwsm)> bestMatchingPeptidesToRemove = new List<(int notch, PeptideWithSetModifications pwsm)>();
-                    List<double> pepValuePredictions = new List<double>();
+                    // Stop loop if canceled
+                    if (GlobalVariables.StopLoops) { return; }
+
+                    ITransformer threadSpecificTrainedModel;
 
-                    //Here we compute the pepvalue predection for each ambiguous peptide in a PSM. Ambiguous peptides with lower pepvalue predictions are removed from the PSM.
-                    foreach (var (Notch, Peptide) in psm.BestMatchingPeptides)
+                    if (outputFolder != null)
+                    {
+                        threadSpecificTrainedModel = mlContext.Model.Load(Path.Combine(outputFolder, "model.zip"), out DataViewSchema savedModelSchema);
+                    }
+                    else
                     {
-                        PsmData pd = CreateOnePsmDataEntry(searchType, fileSpecificParameters, psm, sequenceToPsmCount, fileSpecificTimeDependantHydrophobicityAverageAndDeviation_unmodified, fileSpecificTimeDependantHydrophobicityAverageAndDeviation_modified, fileSpecificMedianFragmentMassErrors, chargeStateMode, Peptide, trainingVariables, Notch, !Peptide.Protein.IsDecoy);
-                        var pepValuePrediction = predictionEngine.Predict(pd);
-                        pepValuePredictions.Add(pepValuePrediction.Probability);
-                        //A score is available using the variable pepvaluePrediction.Score
+                        // single-threaded, prediction model was not saved to hard disk
+                        threadSpecificTrainedModel = trainedModel;
                     }
 
-                    GetIndiciesOfPeptidesToRemove(indiciesOfPeptidesToRemove, pepValuePredictions);
-                    GetBestMatchingPeptidesToRemove(psm, indiciesOfPeptidesToRemove, bestMatchingPeptidesToRemove);
-                    RemoveThePeptides(bestMatchingPeptidesToRemove, psm, pepValuePredictions, ref ambiguousPeptidesRemovedCount);
-                }
-            }
+                    // one prediction engine per thread, because the prediction engine is not thread-safe
+                    var threadPredictionEngine = mlContext.Model.CreatePredictionEngine<PsmData, TruePositivePrediction>(threadSpecificTrainedModel);
+
+                    int ambigousPeptidesRemovedinThread = 0;
+
+                    for (int i = range.Item1; i < range.Item2; i++)
+                    {
+                        PeptideSpectralMatch psm = psms[i];
+
+                        if (psm == null)
+                        {
```

Comment:
nitpicky, but continue has a code smell. please change to if(psm!=null), ditch the continue, and have the brackets cover the rest of the loop.

Reply:
changed now according to your recommendation